### ISSUE TITLE
lagscope reports lower RTT latency in VM (WSL2) than in Windows host if using lagscope as receiver, but vice versa when using latte as receiver 

### ISSUE THREAD

**Issue by wilsonNg007** [microsoft/lagscope#41]
Running two individual experiments:
Running lagscope from WSL2 VM  targeting a remote (windows) device.
Running lagscope from the Windows host (of WSL2) targeting the same remote device.

lagscope reports lower  RTT latency from WSL2 than from the host.  Ran several runs and they all show consistent result.  The expected behavior should have been that the host has lower RTT latency than from the VM.

See attached logs below.
[winHost_lagscope_100Kpings_run2log.txt](https://github.com/microsoft/lagscope/files/6528976/winHost_lagscope_100Kpings_run2log.txt)
[wsl2_lagscope_100Kpings_run2log.txt](https://github.com/microsoft/lagscope/files/6528978/wsl2_lagscope_100Kpings_run2log.txt)

I tried with 1M pings (default) and saw the same result as well.

However, using the ping cmd showing the opposite result i.e. the host reports lower RTT latency than the WSL2 VM, which appears expected behavior.  See below attached.
[winHost_ping_1000pings_32bytes_run0.txt](https://github.com/microsoft/lagscope/files/6528982/winHost_ping_1000pings_32bytes_run0.txt)
[wsl2_ping_1000pings_32bytes_run0.txt](https://github.com/microsoft/lagscope/files/6528983/wsl2_ping_1000pings_32bytes_run0.txt)

**Comment by glima** [comment]
This is most likely a Windows bug, no? Completely different network stacks involved in each case.

**Comment by wilsonNg007** [comment]
WSL2 need to go thru the host via NAT before going out to the wire.

The fact that the previous ping cmd experiment appears showing the expected behavior (i.e. host has lower RTT latency than the VM) indicates that the lagscope may be the highly suspect.

**Comment by glima** [comment]
Yeah, would have to profile it to see what's going on--never touched the code myself.

**Comment by wilsonNg007** [comment]
Hi @santoshx, 

I attached  .csv and console output log from both the host and the VM.   Are these data sufficient to find out why lagscope reports low RTT latency from VM than from the Host ?

<html>
<body>


  | RTT latency (mean) - lagscope
-- | --
Windows Host | 2980.270us
WSA VM | 2814.016us
Difference | -166.254us


</body>
</html>

[winHost_lagscope_console_output_run0.txt](https://github.com/microsoft/lagscope/files/6926049/winHost_lagscope_console_output_run0.txt)
[winHost_lagscope_rawdata_run0.csv](https://github.com/microsoft/lagscope/files/6926053/winHost_lagscope_rawdata_run0.csv)

[wsaVM_lagscope_console_output_run0.txt](https://github.com/microsoft/lagscope/files/6926063/wsaVM_lagscope_console_output_run0.txt)
[wsaVM_lagscope_rawdata_run0.csv](https://github.com/microsoft/lagscope/files/6926064/wsaVM_lagscope_rawdata_run0.csv)

**Comment by wilsonNg007** [comment]
@simonxiaoss , @santoshx,  

I have a new interesting data - If using https://github.com/microsoft/latte running as receiver, I'm getting expected result i.e. window host reported lower RTT latency than in VM. 

lagscope as sender --> lagscope as receiver - issue: VM reports lower RTT latency than Host.
lagscope as sender --> latte as receiver - not repro: VM reports higher RTT latency than Host.

There is an address translation logic between host and VM (whether via NAT or FSE), and that for sure has overhead.  So it makes logical sense for VM to have higher RTT latency than host.

I'll update with the logs.

**Comment by wilsonNg007** [comment]
@santoshx ,  here are the logs from both the Host and VM showing not repro when using latte as receiver.

VM:
[VM_lagscope_console_run0.txt](https://github.com/microsoft/lagscope/files/6940263/VM_lagscope_console_run0.txt)
[VM_lagscope_rawdata_run0.csv](https://github.com/microsoft/lagscope/files/6940265/VM_lagscope_rawdata_run0.csv)
[latteReceiverConsoleOutput_SenderFromVM_run0.txt](https://github.com/microsoft/lagscope/files/6940309/latteReceiverConsoleOutput_SenderFromVM_run0.txt)

Host:
[winHost_lagscope_console_run1.txt](https://github.com/microsoft/lagscope/files/6940280/winHost_lagscope_console_run1.txt)
[winHost_lagscope_rawdata_run1.csv](https://github.com/microsoft/lagscope/files/6940285/winHost_lagscope_rawdata_run1.csv)
[latteReceiverConsoleOutput__SenderFromWinHost_run1.txt](https://github.com/microsoft/lagscope/files/6940319/latteReceiverConsoleOutput__SenderFromWinHost_run1.txt)

**Comment by santoshx** [comment]
I could not repro this on my first attempt. 
I have one Windows laptop as machine 1 and another Windows laptop as machine 2.
Receiver is running on machine 2.
Tried sender from machine 1 directly and sender from machine 1 wsl; both reported same latency.
I'm not concluding that issue isn't there. I'll try few more variations and will get back. 
Thanks.

**Comment by wilsonNg007** [comment]
Is it WSL or WSL2?

WSL and WSL2 are totally different - https://docs.microsoft.com/en-us/windows/wsl/compare-versions  
So it's unlikely you are testing with the right configuration.

FYI, I am using WSA (Windows Subsystem for Android), which is running Microsoft linux kernel, same as WSL2 running microsoft linux kernel.

**Comment by santoshx** [comment]
Seems like I have the WSL1.
C:\>wsl -l -v
  NAME            STATE           VERSION
* Ubuntu-20.04    Running         1

I'll try with WSL2 and get back.
Thanks.
